src/backend/cuda/kernel/ireduce.hpp

```diff
@@ -57,13 +79,16 @@ namespace kernel
         __host__ __device__ MinMaxOp(T val, uint idx) :
             m_val(val), m_idx(idx)
         {
+            if (isNan(val)) {
+                m_val = Binary<T, af_max_t>::init();
+            }
         }
 
         __host__ __device__ void operator()(T val, uint idx)
         {
-            if (cabs(val) > cabs(m_val) ||
-                (cabs(val) == cabs(m_val) &&
-                 idx <= m_idx)) {
```

Comment:
Why was this last condition removed? I think it enforces that if the values are equal then we should prefer indices that are smaller.

Reply:
Another good candidate for a test.

Reply:
I remember that the CUDA reduce and ireduce tests were failing because of that last condition right after I added the isNan condition and the corresponding initial test cases. In any case, this works in a serial reduction, but in CUDA (and I think in a parallel setting in general), that last condition gives incorrect output in the following scenario: suppose that `val > m_val` but `idx > m_idx` - this case is very plausible since we're not going through the array sequentially. In this case, we should assign `m_val` and `m_idx` to `val` and `idx`, but that last condition will prevent that from happening since `idx > m_idx`. The min operation suffered a similar problem. Also I think this is the reason why the OpenCL min/max ireduce kernels don't have this last condition (even before I modified them). Please let me know your thoughts?

Reply:
The (idx) only comes into play when the `val` and `m_val` values are the same (notice that the right-hand side is surrounded by the parenthesis). The new condition should look something like this:
```
(cabs(val) > cabs(m_val) || (cabs(val) == cabs(m_val) && idx <= m_idx)) && !isNan(val)
```

Reply:
OH that makes so much more sense. I think that confused me at first and tried to change the parenthesizing of the conditions, which led me to the failing tests I mentioned earlier. I'll the parenthesizing back (as you've written above) and I'll add the conditions back too. Thanks for the clarification!

Reply:
Actually I'm wrong on another note - the opencl kernel had that last condition also. Will add that back there too.

Reply:
And yep I can add tests for this specific scenario

Reply:
Looking more closely at the original version of the kernels before i modified them, looks like the kernels actually prefer the larger index (the newer element) in case `val == `m_val`. I'll go with that original intent unless you want me to change it to really prefer smaller indices instead?

Reply:
Actually, hold on - that above comment is true for min, but the max is the other way - it prefers the smaller index if available.

Reply:
:+1: just waiting for these changes to pass checks now